pkg/util/log/dedup.go

```diff
@@ -0,0 +1,272 @@
+
+// Deduper implement log.Logger, dedupes log lines.
+type Deduper struct {
+	callerPrefix       string
+	targetKeys         []string
+	maxDedupCount      int
+	expireEntriesAfter time.Duration
+	next               log.Logger
+	stopCh             chan chan struct{}
+	entries            []dedupEntry
+}
+
+// NewDedupLogger creates and initializes a new Deduper instance.
+func NewDedupLogger(callerPrefix string, targetKeys []string, next log.Logger) *Deduper {
+	return newDedupLogger(
+		callerPrefix,
+		targetKeys,
+		next,
+		maxEntries,
+		maxDedupCount,
+		expireEntriesAfter,
+		garbageCollectEvery,
+	)
+}
+
+func newDedupLogger(
+	callerPrefix string,
+	targetKeys []string,
+	next log.Logger,
+	maxEntries int,
+	maxDedupCount int,
+	expireEntriesAfter time.Duration,
+	garbageCollectEvery time.Duration,
+) *Deduper {
+	d := &Deduper{
+		callerPrefix:       callerPrefix,
+		targetKeys:         targetKeys,
+		maxDedupCount:      maxDedupCount,
+		expireEntriesAfter: expireEntriesAfter,
+		next:               next,
+		stopCh:             make(chan chan struct{}, 1),
+		entries:            make([]dedupEntry, maxEntries),
+	}
+	go d.loop(garbageCollectEvery)
+	return d
+}
+
+// Log implements log.Logger.
+func (d *Deduper) Log(keyvals ...interface{}) error {
+	if len(keyvals) == 0 {
+		return nil
+	}
+	if len(keyvals)%2 == 1 {
+		keyvals = append(keyvals, nil)
+	}
+	// check whether log entry matches caller prefix
+	if !d.hasCallerPrefix(keyvals) {
+		return d.next.Log(keyvals...)
+	}
+	// compute dedup key if target keys are present
+	dedupKey := computeDedupKey(keyvals, d.targetKeys)
+	if len(dedupKey) == 0 {
+		return d.next.Log(keyvals...)
+	}
+
+	// look for an active entry
+	for i := 0; i < len(d.entries); i++ {
+		if matched := d.entries[i].updateOrForward(keyvals, dedupKey, d.maxDedupCount, d.next); matched {
+			return nil
+		}
+	}
+	// if no active entry was found, look for an empty slot and register the new entry
+	for i := 0; i < len(d.entries); i++ {
+		if matched := d.entries[i].activate(keyvals, dedupKey); matched {
+			return nil
+		}
+	}
+	// no free slot available... forward entry
+	return d.next.Log(keyvals...)
+}
+
+// Stop stops deduper loop.
+// Invoking this method guarantees all pending entries to be forwarded after returning.
+func (d *Deduper) Stop() {
+	ch := make(chan struct{})
+	d.stopCh <- ch
+	<-ch
+}
+
+func (d *Deduper) loop(collectInterval time.Duration) {
+	tc := time.NewTicker(collectInterval)
+	defer tc.Stop()
+
+	for {
+		select {
+		case <-tc.C:
+			// deactivate expired entries
+			for i := 0; i < len(d.entries); i++ {
+				d.entries[i].deactivateExpired(d.expireEntriesAfter, d.next)
+			}
+
+		case ch := <-d.stopCh:
+			// forward all active entries
+			for i := 0; i < len(d.entries); i++ {
+				d.entries[i].deactivate(d.next)
+			}
+			close(ch) // signal completion
+			return
+		}
+	}
+}
+
+func (d *Deduper) hasCallerPrefix(keyvals []interface{}) bool {
+	valStr, ok := getValue("caller", keyvals).(string)
```

Comment:
The variable name here is `valStr`, but actually `getValue()` returns an `interface{}`. 

Instead of renaming `valStr` I would prefer modifying `getValue()` so it returns a `string`, whenever we can be sure about the type then I think it's better to use the type explicitly instead of `interface{}`.